library/core/src/intrinsics/mir.rs

```diff
@@ -82,8 +228,88 @@ define!("mir_retag_raw", fn RetagRaw<T>(place: T));
 define!("mir_move", fn Move<T>(place: T) -> T);
 define!("mir_static", fn Static<T>(s: T) -> &'static T);
 define!("mir_static_mut", fn StaticMut<T>(s: T) -> *mut T);
+define!(
+    "mir_discriminant",
+    /// Gets the discriminant of a place.
+    ///
+    /// Unfortunately, the "correct" return type of this function depends on the input type in ways
+    /// that cannot be expressed in normal Rust (we would need a fully revealed version of
+    /// `std::mem::Discriminant::<T>`). This function defaults to returning `isize`. That is the
+    /// discriminant type used by rustc on types with no `repr` attribute. However, for types with a
+    /// `repr` attribute, the type may be different. For those purposes you can use
+    /// [`DiscriminantNoInfer`], which allows you to choose the correct return type (getting it
+    /// wrong ICEs in the validator).
+    fn Discriminant<T>(place: T) -> isize
+);
+define!(
+    "mir_discriminant_no_infer",
+    /// Computes the discriminant without pre-selecting the return type.
+    ///
+    /// See [`Discriminant`] for why this is sometimes necessary.
+    fn DiscriminantNoInfer<T, U>(place: T) -> U
+);
```

Comment:
Could we use https://doc.rust-lang.org/std/marker/trait.DiscriminantKind.html#associatedtype.Discriminant here? It's very unstable, but that doesn't matter for custom mir

Reply:
wow, makes me wonder what other kinds of wonderful things that I don't know about are hiding at the back of the standard library. Will update